### PR TITLE
feature(repositorylist): enables drop of folders onto repositorylist

### PR DESCRIPTION
**Pull request by mickedefault**
resolves go-62 

allows dropping folder(s) onto the repository list. If something else is dropped, no action is taken. 

running Add directly after each other didn't trigger OnChange between calls, so AddRange is a work around more than a convenience :)

tried to keep dropadowner as independent as possible but it's still a bit.. static, but I figure if we implement more drop behaviors it will change anyway, so didn't spend to much time on it.

### CODE REVIEW

File: GitOut/Features/Git/RepositoryList/RepositoryListViewModel.cs [hunk 1]
```diff
@@ -81,6 +76,59 @@ ISnackbarService snack
             });
 
             ClearCommand = new CallbackCommand(() => SearchQuery = string.Empty);
+
+            DropCommand = new AsyncCallbackCommand<System.Windows.DataObject>(OnDropAsync);
```
Comment (calahatras) [inline GitOut/Features/Git/RepositoryList/RepositoryListViewModel.cs]: Could use a using instead

File: GitOut/Features/Wpf/DragDrop/DragDropBehavior.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        private static void OnUseAdornerChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        {
+            if (d is not FrameworkElement element)
+            {
+                return;
+            }
+            bool useAdorner = (bool)e.NewValue;
+            if (useAdorner)
+            {
+
```
Comment (calahatras) [inline GitOut/Features/Wpf/DragDrop/DragDropBehavior.cs]: Remove empty line here

File: GitOut/Features/Wpf/DragDrop/DragDropBehavior.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        }
+
+        private static bool HasDirectory(IDataObject dataObject)
+        {
+            if (dataObject is not DataObject data)
+            {
+                return false;
+            }
+
+            if (data.GetFileDropList().Count == 0)
```
Comment (calahatras) [inline GitOut/Features/Wpf/DragDrop/DragDropBehavior.cs]: `data.GetFileDropList` is called multiple times
I propose to use a local variable instead, don't know how "heavy" this call can be

File: GitOut/Features/Wpf/DragDrop/DragDropBehavior.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            {
+                return false;
+            }
+
+            string? firstFile = data.GetFileDropList()[0];
+            if (firstFile == null)
```
Comment (calahatras) [inline GitOut/Features/Wpf/DragDrop/DragDropBehavior.cs]: `is`

File: GitOut/Features/Wpf/DragDrop/DropAdorner.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+        public DropAdorner(UIElement adornedElement, Func<IDataObject, string> adornerText) : base(adornedElement)
+        {
+            IsHitTestVisible = false;
+            adornedElement.DragEnter += OnAdornedElementDragEnter;
+            adornedElement.DragLeave += OnAdornedElementDragLeave;
+            adornedElement.Drop += OnAdornedElementDrop;
+            this.adornerText = adornerText;
+        }
+
```
Comment (calahatras) [inline GitOut/Features/Wpf/DragDrop/DropAdorner.cs:24]: Multiple empty lines

File: GitOut/Features/Wpf/DragDrop/DropAdorner.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+using System.Windows.Documents;
+using System.Windows.Media;
+
+namespace GitOut.Features.Wpf.DragDrop
+{
+    internal class DropAdorner : Adorner
+    {
+        private bool dragging;
+        private string textToRender = string.Empty;
+        private readonly Func<IDataObject, string> adornerText;
```
Comment (calahatras) [inline GitOut/Features/Wpf/DragDrop/DropAdorner.cs]: I prefer having readonly/const variables at top, switch order here

File: GitOut/Features/Wpf/DragDrop/DropAdorner.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+using System.Globalization;
+using System.Windows;
+using System.Windows.Documents;
+using System.Windows.Media;
+
+namespace GitOut.Features.Wpf.DragDrop
+{
+    internal class DropAdorner : Adorner
+    {
+        private bool dragging;
```
Comment (calahatras) [inline GitOut/Features/Wpf/DragDrop/DropAdorner.cs]: use an `is` prefix? `isDragging`

File: GitOut/Features/Wpf/DragDrop/DropAdorner.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            {
+                return;
+            }
+
+            dragging = true;
+            textToRender = adornerText(e.Data);
+
+            base.OnDragEnter(e);
+            InvalidateVisual();
+
```
Comment (calahatras) [inline GitOut/Features/Wpf/DragDrop/DropAdorner.cs]: Remove empty lines

File: GitOut/Features/Wpf/DragDrop/DropAdorner.cs [hunk 9]
```diff
@@ -0,0 +1,86 @@
+using System;
+using System.Globalization;
+using System.Windows;
+using System.Windows.Documents;
+using System.Windows.Media;
+
+namespace GitOut.Features.Wpf.DragDrop
+{
+    internal class DropAdorner : Adorner
```
Comment (calahatras) [inline GitOut/Features/Wpf/DragDrop/DropAdorner.cs]: Is `internal` a slip, should be public imo?

**Review by calahatras — CHANGES_REQUESTED** [review]

File: GitOut/Features/Wpf/DragDrop/DragDropBehavior.cs (hunk 3 above)
Comment (mickedefault) [reply GitOut/Features/Wpf/DragDrop/DragDropBehavior.cs]: > `data.GetFileDropList` is called multiple times I propose to use a local variable instead, don't know how "heavy" this call can be

actually kind of heavy - it creates a new list every time
[source](https://referencesource.microsoft.com/#PresentationCore/Core/CSharp/System/Windows/DataObject.cs,be1f033c907a690b)

File: GitOut/Features/Git/RepositoryList/RepositoryListViewModel.cs (hunk 1 above)
Comment (mickedefault) [reply GitOut/Features/Git/RepositoryList/RepositoryListViewModel.cs]: we're using System.Windows.Forms which also has a DataObject. So then we'd have to differentiate them
would `using DataObject = System.Windows.DataObject;` be better?

Comment (calahatras) [reply GitOut/Features/Git/RepositoryList/RepositoryListViewModel.cs]: I think so, yes. Both would be ok with me, didn't consider `DataObject` being available in another namespace.
Feel free to resolve this comment when you change this.

File: GitOut/Features/Git/RepositoryList/RepositoryListViewModel.cs [hunk 10]
```diff
@@ -81,6 +78,59 @@ ISnackbarService snack
...
+            const string approveText = "YES";
+            SnackAction? action = await snack.ShowAsync(Snack.Builder()
+                .WithMessage($"{Path.GetFileName(path)} is not a valid git repository, do you want to add the folder anyway?")
+                .WithDuration(TimeSpan.FromSeconds(300))
+                .AddAction(approveText));
+
+            return action?.Text == approveText ? repository : null;
+        }
+
+        private async Task OnDropAsync(System.Windows.DataObject? dataObject)
```
Comment (calahatras) [inline GitOut/Features/Git/RepositoryList/RepositoryListViewModel.cs]: Update this as well plz
```suggestion
        private async Task OnDropAsync(DataObject? dataObject)
```

File: GitOut/Features/Git/RepositoryList/RepositoryListViewModel.cs [hunk 11]
```diff
@@ -81,6 +78,59 @@ ISnackbarService snack
...
+        {
+            if (dataObject == null)
```
Comment (calahatras) [inline GitOut/Features/Git/RepositoryList/RepositoryListViewModel.cs]:
```suggestion
            if (dataObject is null)
```

File: GitOut/Features/Wpf/DragDrop/DropAdorner.cs [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+            Brush? brush = DragDropBehavior.GetAdornerStrokeBrush(AdornedElement);
+            Pen renderPen = new(brush, 5) { DashStyle = new DashStyle(new[] { 4d, 4 }, 0) };
+
+            drawingContext.DrawRoundedRectangle(Brushes.Transparent, renderPen, adornedElementRect, 10, 10);
+
+            var textblock = new FormattedText(
+                textToRender,
+                CultureInfo.InvariantCulture,
+                FlowDirection.LeftToRight,
+                new Typeface(new FontFamily("Segoe UI"),
```
Comment (calahatras) [inline GitOut/Features/Wpf/DragDrop/DropAdorner.cs]: Use Roboto

**Review by calahatras — APPROVED** [review]

**Pull request merged**